### ISSUE TITLE
Revisit uniqueness of usernames in CoG

### ISSUE THREAD

**Issue by LucaCinquini** [EarthSystemCoG/COG#1221]
Who: Prashanth

CoG doesn't allow local user creation if same username from a different openid provider has logged into CoG

**Comment by LucaCinquini** [comment]
It turns out that Django REQUIRES usernames to be unique in a database. See extracts from the schema dump:

CREATE TABLE auth_user (
    id integer NOT NULL,
    password character varying(128) NOT NULL,
    last_login timestamp with time zone,
    is_superuser boolean NOT NULL,
    username character varying(30) NOT NULL,
    first_name character varying(30) NOT NULL,
    last_name character varying(30) NOT NULL,
    email character varying(254) NOT NULL,
    is_staff boolean NOT NULL,
    is_active boolean NOT NULL,
    date_joined timestamp with time zone NOT NULL
);

ALTER TABLE ONLY auth_user
    ADD CONSTRAINT auth_user_username_key UNIQUE (username);

So we cannot change this constraint without much effort and the likely possibility of introducing bugs.

**Comment by LucaCinquini** [comment]
But what we can do is the following: when a user from another IdP logs onto CoG, he/she will be assigned a username that is NOT exactly equal to the last part of their openid, but rather that part + an incremental number.
For example, if I log onto my local CoG with openid:

https://pcmdi.llnl.gov/esgf-idp/openid/lucacinquini

I will be assigned a local username of "lucacinquini1". This way, "lucacinquini" is still available for local users if they want to create such an account.

**Comment by pchengi** [comment]
As long as CoG doesn't disallow registering with the same name (solved by doing auto increments), that's fine. This is a perfectly acceptable solution.

**Comment by LucaCinquini** [comment]
Hi Prashanth,
	actually that’s the point: if there is already a username=“lucacinquini1” in the database (generated because lucacinquini from DKRZ logged onto that CoG), CoG cannot
create an account with username=“lucacinquini1” (django will NOT allow it). But the user can choose “username=lucacinquini” or “username=lucacinquini2” etc..

thanks, L

**Comment by pchengi** [comment]
Luca, I meant, it doesn't matter what value is stored in the CoG database, but the openid should have the correct name. If a user pchengi is already registered from a different node, I don't mind if CoG stores the user information as pchengi1 locally, but the local openid should NOT end as pchengi1, but pchengi. Hope I'm being clear.

Regards
Prashanth

**Comment by LucaCinquini** [comment]
The functionality requested by Prashanth has been implemented. The account creation workflow has changed. Now, if a user requests username="xyzuser", CoG checks weather there is a local openid such as: https://hostnmame/esgf-idp/openid/xyzuser in the database. If yes, it returns an error - the user will have to choose a different username. if not, it considers the openid available and it will create it in the ESGCET database. Them, if possible, it assigned the user the username "xyzuser". If that is not available, it tries xyzuser1, xyzuser2, etc....
In other words: availability of the requested openid (not the username) in the local database is the critical factor when creating an account.

**Comment by murphysj** [comment]
@LucaCinquini 
I can't test this since cu-dev nor my local sandbox has an IdP. Close this if you have tested locally.

**Comment by LucaCinquini** [comment]
Tested at JPL after I deployed the RC there.
